spec.html

```diff
@@ -28955,51 +28955,71 @@ <h1>
-                1. Else,
-                  1. Let _S_ be the substring of _string_ from _start_ to _k_ + 1.
+                1. Let _c_ be the code unit whose numeric value is _B_.
+                1. If _c_ is in _preserveEscapeSet_, set _S_ to _escape_. Otherwise, set _S_ to _c_.
               1. Else,
                 1. If _n_ = 1 or _n_ > 4, throw a *URIError* exception.
-                1. If _k_ + (3 × (_n_ - 1)) ≥ _strLen_, throw a *URIError* exception.
                 1. Let _Octets_ be « _B_ ».
                 1. Let _j_ be 1.
                 1. Repeat, while _j_ &lt; _n_,
                   1. Set _k_ to _k_ + 1.
+                  1. If _k_ + 3 > _len_, throw a *URIError* exception.
                   1. If the code unit at index _k_ within _string_ is not the code unit 0x0025 (PERCENT SIGN), throw a *URIError* exception.
-                  1. If the code units at index (_k_ + 1) and (_k_ + 2) within _string_ do not represent hexadecimal digits, throw a *URIError* exception.
-                  1. Let _B_ be the 8-bit value represented by the two hexadecimal digits at index (_k_ + 1) and (_k_ + 2).
+                  1. Let _continuationByte_ be ParseHexOctet(_string_, _k_ + 1).
+                  1. If _continuationByte_ is not an integer, throw a *URIError* exception.
+                  1. Append _continuationByte_ to _Octets_.
                   1. Set _k_ to _k_ + 2.
-                  1. Append _B_ to _Octets_.
                   1. Set _j_ to _j_ + 1.
                 1. Assert: The length of _Octets_ is _n_.
                 1. If _Octets_ does not contain a valid UTF-8 encoding of a Unicode code point, throw a *URIError* exception.
                 1. Let _V_ be the code point obtained by applying the UTF-8 transformation to _Octets_, that is, from a List of octets into a 21-bit value.
-                1. Let _S_ be UTF16EncodeCodePoint(_V_).
+                1. Set _S_ to UTF16EncodeCodePoint(_V_).
             1. Set _R_ to the string-concatenation of _R_ and _S_.
             1. Set _k_ to _k_ + 1.
+          1. Return _R_.
         </emu-alg>
         <emu-note>
           <p>RFC 3629 prohibits the decoding of invalid UTF-8 octet sequences. For example, the invalid sequence 0xC0 0x80 must not decode into the code unit 0x0000. Implementations of the Decode algorithm are required to throw a *URIError* when encountering such invalid sequences.</p>
         </emu-note>
       </emu-clause>
+
+      <emu-clause id="sec-parsehexoctet" type="abstract operation">
+        <h1>
+          ParseHexOctet (
+            _string_: a String,
+            _position_: a non-negative integer,
+          ): either a non-negative integer or a non-empty List of *SyntaxError* objects
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>It parses a sequence of two hexadecimal characters at the specified _position_ in _string_ into an unsigned 8-bit integer.</dd>
```

Comment:
```suggestion
          <dd>It parses a sequence of two hexadecimal code units at the specified _position_ in _string_ into an unsigned 8-bit integer.</dd>
```
We mostly avoid using the word "character" outside of RegExps.